### PR TITLE
V2 : additional tests for modeling

### PR DESCRIPTION
**Pull request by z3dev**
These changes include additional test suites for exposed functionality, increasing coverage.
- mat4.isMirroring()
- vec4.dot()
- measurement.measureEpsilon()
- slice.equals()

The following functions and tests suites were dropped.
- poly3.equals() : incorrect comparisons, unused
- geom3.equals() : incorrect comparisons, unused

Documentation for these functions was improved as well.

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

File: packages/modeling/src/operations/extrusions/slice/equals.test.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const test = require('ava')
+
+const { equals, fromPoints } = require('./index')
+
+test('slice: equals() should return proper value', (t) => {
+  const sliceA = fromPoints([[0, 0], [1, 0], [1, 1]])
+  const sliceB = fromPoints([[0, 1], [1, 0], [1, 1]])
+  const sliceC = fromPoints([[0, 0], [1, 0], [1, 1], [0, 0]])
+
+  t.is(equals(sliceA, sliceA), true)
```
Comment (SimonClark) [inline packages/modeling/src/operations/extrusions/slice/equals.test.js:10]: Ditto on the equality test

File: packages/modeling/src/operations/extrusions/slice/equals.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ const vec3 = require('../../../maths/vec3')
  * Determine if the given slices are equal.
```
Comment (SimonClark) [inline packages/modeling/src/operations/extrusions/slice/equals.js:4]: Maybe a better definition of what "equal" means here.

File: packages/modeling/src/geometries/geom3/equals.test.js [hunk 3]
```diff
@@ -1,7 +1,17 @@
...
 const { equals, fromPoints } = require('./index')
 
 test('equals: checks if two geom3 are equal', (t) => {
-  const points = [[[0, 0, 0], [1, 0, 0], [1, 0, 1]]]
-  t.is(equals(fromPoints(points), fromPoints(points)), true)
+  const geomA = fromPoints([[[0, 0, 0], [1, 0, 0], [1, 0, 1]]])
+  const geomB = fromPoints([[[1, 1, 1], [1, 0, 0], [1, 0, 1]]])
+  const geomC = fromPoints([[[0, 0, 0], [1, 0, 0], [1, 0, 1]], [[1, 1, 1], [1, 1, 1], [1, 1, 1]]])
+
+  t.is(equals(geomA, geomA), true)
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometries/geom3/equals.test.js:10]: Actually, thinking about this. equals() is probably not possible as the "order" of polygons is NOT guaranteed, as well as the "order" of points in each polygon is NOT guaranteed.

How about dropping equals()?

File: packages/modeling/src/geometries/poly3/equals.test.js [hunk 4]
```diff
@@ -1,7 +1,17 @@
 const test = require('ava')
+
 const { equals, fromPoints } = require('./index')
 
 test('equals: checks if two poly3 are equal', (t) => {
-  const points = [[0, 0, 0], [1, 0, 0], [1, 0, 1]]
-  return t.is(equals(fromPoints(points), fromPoints(points)), true)
+  const polyA = fromPoints([[0, 0, 0], [1, 0, 0], [1, 0, 1]])
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometries/poly3/equals.test.js:6]: See comment above.

If equals is required then the comparison of points needs to disregard "order".

File: packages/modeling/src/maths/mat4/isMirroring.test.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  const planeZ = plane.fromPoints([0, 0, 0], [1, 0, 0], [1, 1, 0])
+
+  matrix = mirrorByPlane(planeX)
+  t.true(isMirroring(matrix))
+
+  matrix = mirrorByPlane(planeY)
+  t.true(isMirroring(matrix))
+
+  matrix = mirrorByPlane(planeZ)
+  t.true(isMirroring(matrix))
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/maths/mat4/isMirroring.test.js:27]: See lines 11 and 14.

File: packages/modeling/src/operations/extrusions/slice/equals.test.js (hunk 1 above)
Comment (z3dev) [reply packages/modeling/src/operations/extrusions/slice/equals.test.js:10]: Same comment as above. There is no "order" of edges, so the logic needs to change or this function should be dropped.

File: packages/modeling/src/geometries/geom3/equals.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonClark) [reply packages/modeling/src/geometries/geom3/equals.test.js:10]: it's not terribly useful for comparisons as it is, no.  And I don't see much value to it in common modelling workflows.

File: packages/modeling/src/maths/mat4/isMirroring.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (SimonClark) [reply packages/modeling/src/maths/mat4/isMirroring.test.js:27]: Yep, those test identity and scaling matrix.  What I'm asking. if I have a matrix that mirrors and scales, is/should isMirroring() return true?

File: packages/modeling/src/geometries/geom3/equals.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometries/geom3/equals.test.js]: Removed geom3.equals and test suite.

File: packages/modeling/src/geometries/geom3/equals.test.js [hunk 6]
```diff
@@ -1,7 +1,17 @@
...
+
+  t.is(equals(geomA, geomB), false)
```
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometries/geom3/equals.test.js]: Removed geom3.equals and test suite

File: packages/modeling/src/operations/extrusions/slice/equals.test.js (hunk 1 above)
Comment (z3dev) [reply packages/modeling/src/operations/extrusions/slice/equals.test.js]: Actually, this is being used by extrudeBySlices to determine if initial and final slices are equal, and adding another segment if not equal.

File: packages/modeling/src/operations/extrusions/slice/equals.js (hunk 2 above)
Comment (z3dev) [reply packages/modeling/src/operations/extrusions/slice/equals.js]: Done

File: packages/modeling/src/geometries/poly3/equals.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/geometries/poly3/equals.test.js]: Removed poly3.equals and test suite

File: packages/modeling/src/operations/extrusions/slice/equals.test.js (hunk 1 above)
Comment (z3dev) [reply packages/modeling/src/operations/extrusions/slice/equals.test.js]: Tests have been simplified.

File: packages/modeling/src/maths/mat4/isMirroring.test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (z3dev) [reply packages/modeling/src/maths/mat4/isMirroring.test.js:27]: Added more test cases, compouding tranforms (rotate, scale, translate).

**Review by SimonClark — APPROVED** [review]

**Pull request merged**